File: app/src/main/java/com/galtashma/parsedashboard/ParseServerConfigStorage.java
package com.galtashma.parsedashboard;

import android.content.Context;
import android.content.SharedPreferences;
import com.afollestad.ason.AsonArray;
import java.util.List;

/**
 * Created by gal on 3/16/18.
 */

public class ParseServerConfigStorage {

    private static final String PREF_KEY = "parse_server_key";
    private static final String PREF_SERVERS_KEY = "parse_server_config_key";

    private final Context context;

    public ParseServerConfigStorage(Context context){
        this.context = context;
    }

    public  void saveServer(ParseServerConfig config){
        AsonArray<ParseServerConfig> servers = getServersAson();
        servers.add(config);

        SharedPreferences pref = context.getSharedPreferences(PREF_KEY, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putString(PREF_SERVERS_KEY, servers.toString());
        editor.commit();
    }

    private AsonArray<ParseServerConfig> getServersAson(){
        SharedPreferences pref = context.getSharedPreferences(PREF_KEY, Context.MODE_PRIVATE);
        String input = pref.getString(PREF_SERVERS_KEY, "[]");

        AsonArray<ParseServerConfig> array = new AsonArray<>(input);


        return array;
    }

    public List<ParseServerConfig> getServers(){
        AsonArray<ParseServerConfig> servers = getServersAson();
        return servers.deserializeList(ParseServerConfig.class);
    }

}

File: app/src/main/java/com/galtashma/parsedashboard/ParseServerConfigStorage.kt
package com.galtashma.parsedashboard

import android.content.Context
import com.afollestad.ason.AsonArray

/**
 * Created by gal on 3/16/18, rewritten by Cyb3rKo on 05/12/22.
 */

private const val PREF_KEY = "parse_server_key"
private const val PREF_SERVERS_KEY = "parse_server_config_key"

class ParseServerConfigStorage(val context: Context) {
    fun saveServer(config: ParseServerConfig) {
        val servers = getServersAson()
        servers.add(config)
        overrideServersAson(servers)
    }

    fun deleteServer(appId: String) {
        val servers = getServers()
        var toRemove: ParseServerConfig? = null

        servers.forEach {
            if (it.appId == appId) {
                toRemove = it
            }
        }

        if (toRemove != null) {
            servers.remove(toRemove)
            overrideServers(servers)
        }
    }

    fun getServers(): MutableList<ParseServerConfig> {
        val servers = getServersAson()
        return servers.deserializeList(ParseServerConfig::class.java)
    }

    private fun getServersAson(): AsonArray<ParseServerConfig> {
        val pref = context.getSharedPreferences(PREF_KEY, Context.MODE_PRIVATE)
        val input = pref.getString(PREF_SERVERS_KEY, "[]")
        return AsonArray(input)
    }

    private fun overrideServers(servers: List<ParseServerConfig>) {
        val asonArray = AsonArray<ParseServerConfig>()
        servers.forEach {
            asonArray.add(it)
        }

        overrideServersAson(asonArray)
    }

    private fun overrideServersAson(servers: AsonArray<ParseServerConfig>) {
        val pref = context.getSharedPreferences(PREF_KEY, Context.MODE_PRIVATE)
        val editor = pref.edit()
        editor.putString(PREF_SERVERS_KEY, servers.toString())
        editor.apply()
    }
}
